libbpf-tools/memleak.c

```diff
@@ -668,53 +787,7 @@ int print_stack_frames(struct allocation *allocs, size_t nr_allocs, int stack_tr
 			return -errno;
 		}
 
-#ifdef USE_BLAZESYM
-		const blazesym_result *result = blazesym_symbolize(symbolizer, &src_cfg, 1, stack, env.perf_max_stack_depth);
-
-		for (size_t j = 0; j < result->size; ++j) {
-			const uint64_t addr = stack[j];
-
-			if (addr == 0)
-				break;
-
-			// no symbol found
-			if (!result || j >= result->size || result->entries[j].size == 0) {
-				print_stack_frame(j, addr, NULL);
-
-				continue;
-			}
-
-			// single symbol found
-			if (result->entries[j].size == 1) {
-				const blazesym_csym *sym = &result->entries[j].syms[0];
-				print_stack_frame(j, addr, sym);
-
-				continue;
-			}
-
-			// multi symbol found
-			printf("\t%zu [<%016lx>] (%lu entries)\n", j, addr, result->entries[j].size);
-
-			for (size_t k = 0; k < result->entries[j].size; ++k) {
-				const blazesym_csym *sym = &result->entries[j].syms[k];
-				if (sym->path && strlen(sym->path))
-					printf("\t\t%s@0x%lx %s:%ld\n", sym->symbol, sym->start_address, sym->path, sym->line_no);
-				else
-					printf("\t\t%s@0x%lx\n", sym->symbol, sym->start_address);
-			}
-		}
-
-		blazesym_result_free(result);
-#else
-		for (size_t j = 0; j < env.perf_max_stack_depth; ++j) {
-			const uint64_t addr = stack[j];
-
-			if (addr == 0)
-				break;
-
-			printf("\t%zu [<%016lx>]\n", j, addr);
-		}
-#endif
+		(*print_stack_frames_func)();
```

Comment:
I wonder if we can validate the existence of this function somewhere early on. there's a comment "post-processing and validation of env settings" where checks are done on env values

Reply:
Hi @inwardvessel 

Thanks for the comments.

I assume that the place you mentioned is too early ? `print_stack_frames_func` is determined by whether the initialization of various `symbolizers` is successful at runtime.
And based on the current implementation, `print_stack_frames_func` should always be valid, otherwise, the program would have jumped to `cleanup` already.

Reply:
fair enough